logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/ConversationVerificationStatusHandler.kt

```diff
@@ -18,45 +18,152 @@
+ * Observes Conversation Verification status.
+ * Notify user (by adding System message in conversation) if needed about changes.
+ * @param conversationId [ConversationId]
+ * @param shouldShareFlow [Boolean] flag if the resulted [Flow] should be shared in IO [CoroutineScope], or not.
+ * Basically needed only for unit-tests cause it fails if flow is shared (Flow never completes).
  */
 internal interface ConversationVerificationStatusHandler {
-    suspend operator fun invoke(conversation: Conversation, status: ConversationVerificationStatus)
+    suspend operator fun invoke(conversationId: ConversationId, shouldShareFlow: Boolean = true): Flow<Unit>
 }
 
 internal class ConversationVerificationStatusHandlerImpl(
     private val conversationRepository: ConversationRepository,
     private val persistMessage: PersistMessageUseCase,
+    private val mlsConversationRepository: MLSConversationRepository,
     private val selfUserId: UserId,
     kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl
 ) : ConversationVerificationStatusHandler {
     private val dispatcher = kaliumDispatcher.io
 
-    override suspend fun invoke(conversation: Conversation, status: ConversationVerificationStatus): Unit = withContext(dispatcher) {
-        if (shouldNotifyUser(conversation, status)) {
-            val content = when (conversation.protocol) {
+    private val lock = Mutex()
+    private val observeFlows = hashMapOf<ConversationId, Flow<Unit>>()
+
+    override suspend fun invoke(conversationId: ConversationId, shouldShareFlow: Boolean): Flow<Unit> = withContext(dispatcher) {
+        lock.withLock {
+            val currentFlow = observeFlows[conversationId]
+            if (currentFlow != null) return@withContext currentFlow
+        }
+
+        val flow = conversationRepository.getConversationProtocolInfo(conversationId)
+            .map { protocol ->
+                fetchVerificationStatusFlow(conversationId, protocol)
+                    .mapLatest { status -> notifyUserIfNeeded(conversationId, protocol, status) }
+                    .onCompletion { observeFlows.remove(conversationId) }
+            }
+            .getOrElse(emptyFlow())
+            .apply {
+                if (shouldShareFlow) shareIn(this@withContext, SharingStarted.WhileSubscribed(1))
+            }
+
+        lock.withLock { observeFlows[conversationId] = flow }
+
+        flow
+    }
+
+    /**
+     * Get conversation verification status and save it locally.
+     */
+    private suspend fun fetchVerificationStatusFlow(
+        conversationId: ConversationId,
+        protocol: Conversation.ProtocolInfo
+    ): Flow<VerificationStatus> =
+        observerVerificationStatus(protocol, conversationId)
+            .distinctUntilChanged()
+            .onlyRight()
+            .mapLatest { newStatus ->
+                val currentStatus = conversationRepository.getConversationVerificationStatus(conversationId)
```

Comment:
question: do we need to call CC anymore? since we update/fetch the state based on the epoch changes? here simply can't we only observe the value in the db?

Reply:
I'm not sure if I got your question. 
We observe epoch changes, but it only triggers us that something was changed and we need to check if the Verification Status changed (by calling CC).